### PR TITLE
dealii: add cuda support

### PR DESCRIPTION
**Pull request by BenBrands**
@davydden @jppelteret ping.

### CODE REVIEW

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 1]
```diff
@@ -212,6 +217,18 @@ def cmake_args(self):
                 '-DDEAL_II_DOCHTML_RELDIR=share/deal.II/doc'
             ])
 
+        # CUDA
+        # FIXME  -DDEAL_II_CUDA_FLAGS="-arch=sm_60"
```
Comment (davydden) [inline var/spack/repos/builtin/packages/dealii/package.py]: @tgamblin i presume so far there is nothing in Spack which would help to know the CUDA device architecture?

**Comment by Rombur** [comment]
@davydden how do you run the tests when you install deal.II with spack? I was able to build deal with cuda support (thanks @BenBrands) but how do I run the tests?

**Comment by davydden** [comment]
@Rombur 

> @davydden how do you run the tests when you install deal.II with spack?

essentially, there are two ways:
 1. keep the build prefix `spack install --keep-prefix dealii` and run manually the test-suite as usual.
2. setup deal.II tests for an already installed library 
```
$ mkdir tests_for_installed_dealii
$ cd tests_for_installed_dealii
$ cmake -DDEAL_II_DIR=$(spack location -i dealii) /path/to/dealii_source/tests
```
(here i assumed you have a single installation of dealii in spack, otherwise you can be more precise in specs for location, i.e. `spack location -i dealii@develop`)

**Comment by Rombur** [comment]
@davydden thanks!

**Comment by davydden** [comment]
@BarrySmith suggested that `spack` should have a way to run large test-suite post-installation (which will also be very useful for dealii), see https://github.com/LLNL/spack/issues/2060 . But unfortunately no one yet implemented this.

**Comment by davydden** [comment]
@Rombur it should be ` --keep-stage` instead, sorry, mixed the two.

**Comment by Rombur** [comment]
I tried the second method but I get this error:
` The C++ compiler "/home/spack/lib/spack/env/gcc/g++" is not able to compile
  a simple test program.`
 `Spack compiler must be run from Spack! Input 'SPACK_PREFIX' is missing.`

**Comment by davydden** [comment]
@Rombur i will look into that, this path is a path to compiler wrappers which indeed should be run directly from Spack only. Unless we picked them um somehow during deal.II config, we shall never see them. For example `openmpi` and alike have a step where they [filter those compiler wrappers and substitute by actual paths to the compilers](https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/openmpi/package.py#L368-L407). Not sure how/why this happens here... Last time I tried running the complete test suite was several months ago.

**Comment by Rombur** [comment]
Same problem with the first method. Anyway the PR looks good and it compiles fine.

**Comment by davydden** [comment]
@alalazo : the PR is approved by 3 devs, I think we are good to go, if you have a second, please hit the green button :smile:

**Comment by davydden** [comment]
> @davydden I trust your dev word on this! 😄

:smile: thanks!

**Comment by davydden** [comment]
@Rombur i was re-installing everything on my mac due to a new version of clang and don't see Spack's compile wrappers hard-coded in the installation of dealii:
```
$ ag "lib/spack/env" /Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/dealii-develop-rixvq6nkgcckw37plkmfyiyrkaxmebsv
```
what we have is MPI wrappers:
```
SET(DEAL_II_CXX_COMPILER "/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/openmpi-3.0.0-o7lxsc2dpscn6ydhv6itqhly3k7i4r2m/bin/mpic++")
SET(DEAL_II_C_COMPILER "/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/openmpi-3.0.0-o7lxsc2dpscn6ydhv6itqhly3k7i4r2m/bin/mpicc")
```
which appears to be ok as well:
```
$ ag "lib/spack/env" /Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-9.0.0-apple/openmpi-3.0.0-o7lxsc2dpscn6ydhv6itqhly3k7i4r2m
```
Setting up tests worked for me. Maybe the whole problem is there only when `cuda` is enabled? Maybe Spack has to have a post-installation step to filter out its compiler wrappers from `nvcc` and alike, similar to https://github.com/LLNL/spack/blob/develop/var/spack/repos/builtin/packages/openmpi/package.py#L369-L408
Could you please check if compiler wrappers appear in your CUDA installation within Spack? If won't have time, i can check it on Monday with @BenBrands

**Comment by BenBrands** [comment]
@Rombur CUDA works fine for us
```
$ ctest -R "cuda"
Test project /home/benjamin/dealii_tests/_build
      Start  1: cuda/cuda_evaluate_1d_shape.debug
 1/14 Test  #1: cuda/cuda_evaluate_1d_shape.debug ...........   Passed    0.53 sec
      Start  2: cuda/cuda_evaluate_1d_shape.release
 2/14 Test  #2: cuda/cuda_evaluate_1d_shape.release .........   Passed   28.26 sec
      Start  3: cuda/cuda_evaluate_2d_shape.debug
 3/14 Test  #3: cuda/cuda_evaluate_2d_shape.debug ...........   Passed   23.69 sec
      Start  4: cuda/cuda_evaluate_2d_shape.release
 4/14 Test  #4: cuda/cuda_evaluate_2d_shape.release .........   Passed   21.66 sec
      Start  5: cuda/cuda_tensor_01.debug
 5/14 Test  #5: cuda/cuda_tensor_01.debug ...................   Passed   11.75 sec
      Start  6: cuda/cuda_tensor_01.release
 6/14 Test  #6: cuda/cuda_tensor_01.release .................   Passed   11.09 sec
      Start  7: cuda/cuda_vector_01.debug
 7/14 Test  #7: cuda/cuda_vector_01.debug ...................   Passed    9.28 sec
      Start  8: cuda/cuda_vector_01.release
 8/14 Test  #8: cuda/cuda_vector_01.release .................   Passed    7.50 sec
      Start  9: cuda/cuda_vector_02.debug
 9/14 Test  #9: cuda/cuda_vector_02.debug ...................   Passed    7.28 sec
      Start 10: cuda/cuda_vector_02.release
10/14 Test #10: cuda/cuda_vector_02.release .................   Passed    8.75 sec
      Start 11: cuda/matrix_free_matrix_vector_01.debug
11/14 Test #11: cuda/matrix_free_matrix_vector_01.debug .....   Passed   43.36 sec
      Start 12: cuda/matrix_free_matrix_vector_01.release
12/14 Test #12: cuda/matrix_free_matrix_vector_01.release ...   Passed   35.14 sec
      Start 13: cuda/matrix_free_matrix_vector_04.debug
13/14 Test #13: cuda/matrix_free_matrix_vector_04.debug .....   Passed   41.98 sec
      Start 14: cuda/matrix_free_matrix_vector_04.release
14/14 Test #14: cuda/matrix_free_matrix_vector_04.release ...   Passed   39.20 sec

100% tests passed, 0 tests failed out of 14

Total Test time (real) = 295.41 sec
```

**Comment by Rombur** [comment]
@BenBrands OK thanks. At the time I was working on a container where I was root, maybe that's why I had problems.

**Pull request merged**